internal/concierge/impersonator/impersonator.go

```diff
@@ -0,0 +1,237 @@
+	}
+
+	kubeTransportConfig, err := kubeconfig.TransportConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster transport config: %w", err)
+	}
+	kubeTransportConfig.TLS.NextProtos = []string{"http/1.1"}
+
+	kubeRoundTripper, err := transport.New(kubeTransportConfig)
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster transport: %w", err)
+	}
+
+	reverseProxy := httputil.NewSingleHostReverseProxy(serverURL)
+	reverseProxy.Transport = kubeRoundTripper
+
+	return &proxy{
+		cache:       cache,
+		jsonDecoder: jsonDecoder,
+		proxy:       reverseProxy,
+		log:         log,
+	}, nil
+}
+
+func (p *proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	log := p.log.WithValues(
+		"url", r.URL.String(),
+		"method", r.Method,
+	)
+
+	if err := ensureNoImpersonationHeaders(r); err != nil {
+		log.Error(err, "impersonation header already exists")
+		http.Error(w, "impersonation header already exists", http.StatusBadRequest)
+		return
+	}
+
+	// Never mutate request (see http.Handler docs).
+	newR := r.Clone(r.Context())
+
+	authentication, authenticated, err := bearertoken.New(authenticator.TokenFunc(func(ctx context.Context, token string) (*authenticator.Response, bool, error) {
+		tokenCredentialReq, err := extractToken(token, p.jsonDecoder)
+		if err != nil {
+			log.Error(err, "invalid token encoding")
+			return nil, false, &httpError{message: "invalid token encoding", code: http.StatusBadRequest}
+		}
+
+		log = log.WithValues(
+			"authenticator", tokenCredentialReq.Spec.Authenticator,
+		)
+
+		userInfo, err := p.cache.AuthenticateTokenCredentialRequest(newR.Context(), tokenCredentialReq)
+		if err != nil {
+			log.Error(err, "received invalid token")
+			return nil, false, &httpError{message: "invalid token", code: http.StatusUnauthorized}
+		}
+		if userInfo == nil {
+			log.Info("received token that did not authenticate")
+			return nil, false, &httpError{message: "not authenticated", code: http.StatusUnauthorized}
+		}
+		log = log.WithValues("userID", userInfo.GetUID())
+
+		return &authenticator.Response{User: userInfo}, true, nil
+	})).AuthenticateRequest(newR)
+	if err != nil {
+		httpErr, ok := err.(*httpError)
+		if !ok {
+			log.Error(err, "unrecognized error")
+			http.Error(w, "unrecognized error", http.StatusInternalServerError)
+		}
+		http.Error(w, httpErr.message, httpErr.code)
+		return
+	}
+	if !authenticated {
+		log.Error(constable.Error("token authenticator did not find token"), "invalid token encoding")
+		http.Error(w, "invalid token encoding", http.StatusBadRequest)
+		return
+	}
+
+	newR.Header = getProxyHeaders(authentication.User, r.Header)
+
+	log.Info("proxying authenticated request")
+	p.proxy.ServeHTTP(w, newR)
+}
+
+type httpError struct {
+	message string
+	code    int
+}
+
+func (e *httpError) Error() string { return e.message }
+
+func ensureNoImpersonationHeaders(r *http.Request) error {
```

Comment:
This should check for the Impersonate prefix on any header since Kube could add more.